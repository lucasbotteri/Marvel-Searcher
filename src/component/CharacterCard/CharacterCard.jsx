import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Star from "../Star";

const CharacterCardWrapper = styled.section`
  cursor: pointer;
  height: 380px;
  position: relative;
  width: 256px;
`;

const ImageOverlay = styled.img`
  border-radius: ${(props) => props.theme.radius};
  height: 100%;
  position: relative;
  width: 100%;
`;

// TODO color should be themed, fontSize also
const Name = styled.h2`
  bottom: 8px;
  color: ${(props) => props.theme.primaryColor};
  font-size: ${(props) => props.theme.h2Size};
  left: 16px;
  position: absolute;
`;

const TopRightStar = styled(Star)`
  position: absolute;
  right: 16px;
  top: 8px;
`;
const CharacterCard = ({ character, onClick, isFaved }) => {
  const { name, thumbnail } = character;
  return (
    <CharacterCardWrapper onClick={onClick} alt={name}>
      <ImageOverlay src={thumbnail} />
      <TopRightStar isSolid={isFaved} />
      <Name> {name} </Name>
    </CharacterCardWrapper>
  );
};

CharacterCard.propTypes = {
  character: PropTypes.shape({
    name: PropTypes.string,
    thumbnail: PropTypes.string,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
  isFaved: PropTypes.bool,
};

CharacterCard.defaultProps = {
  isFaved: false,
};

export default CharacterCard;
