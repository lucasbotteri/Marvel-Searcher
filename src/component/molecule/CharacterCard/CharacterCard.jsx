import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Star from "../Star";

export const CharacterCardWrapper = styled.section`
  cursor: pointer;
  height: 380px;
  position: relative;
  width: 256px;

  &:before {
    content: "";
    background: linear-gradient(
      to top,
      rgba(0, 0, 0, 1) 0%,
      rgba(255, 255, 255, 0) 33%
    );
    border-radius: ${(props) => props.theme.radius};
    bottom: 0;
    left: 0;
    position: absolute;
    right: 0;
    top: 0;
    z-index: 1;
  }
`;

export const ImageOverlay = styled.img`
  background: grey;
  border-radius: ${(props) => props.theme.radius};
  height: 100%;
  position: relative;
  width: 100%;
`;

export const Name = styled.h2`
  bottom: 8px;
  color: ${(props) => props.theme.primaryColor};
  font-size: ${(props) => props.theme.h2Size};
  left: 16px;
  position: absolute;
  z-index: 1;
`;

export const TopRightStar = styled(Star)`
  position: absolute;
  right: 16px;
  top: 8px;
`;

const CharacterCard = ({ character, onClick, isFaved, onFaved, onUnfaved }) => {
  const { name, thumbnail } = character;
  return (
    <CharacterCardWrapper onClick={onClick} alt={name}>
      <ImageOverlay src={thumbnail} />
      <TopRightStar
        isSolid={isFaved}
        onClick={(e) => {
          e.stopPropagation();
          if (isFaved) {
            onUnfaved();
          } else {
            onFaved();
          }
        }}
      />
      <Name> {name} </Name>
    </CharacterCardWrapper>
  );
};

CharacterCard.propTypes = {
  character: PropTypes.shape({
    name: PropTypes.string,
    thumbnail: PropTypes.string,
  }).isRequired,
  onClick: PropTypes.func.isRequired,
  onFaved: PropTypes.func.isRequired,
  onUnfaved: PropTypes.func.isRequired,
  isFaved: PropTypes.bool,
};

CharacterCard.defaultProps = {
  isFaved: false,
};

export default CharacterCard;
