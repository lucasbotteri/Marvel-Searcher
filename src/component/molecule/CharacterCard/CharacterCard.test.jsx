import React from "react";
import { shallow } from "enzyme";
import CharacterCard, { TopRightStar } from "./CharacterCard";

const character = {
  id: 1014997,
  name: "Deathstrike (Ultimate)",
  thumbnail: "someurl.com",
};

describe("CharacterCard", () => {
  it("Should render without crashing", () => {
    const wrapper = shallow(
      <CharacterCard
        character={character}
        isFaved={false}
        onClick={() => {}}
        onFaved={() => {}}
        onUnfaved={() => {}}
      />
    );
    expect(wrapper).toBeTruthy();
  });

  it("Should show solid star", () => {
    const wrapper = shallow(
      <CharacterCard
        character={character}
        onClick={() => {}}
        onFaved={() => {}}
        onUnfaved={() => {}}
        isFaved
      />
    );

    expect(wrapper.find(TopRightStar).prop("isSolid")).toBeTruthy();
  });

  it("Should show not solid star", () => {
    const wrapper = shallow(
      <CharacterCard
        character={character}
        isFaved={false}
        onClick={() => {}}
        onFaved={() => {}}
        onUnfaved={() => {}}
      />
    );
    expect(wrapper.find(TopRightStar).prop("isSolid")).toBeFalsy();
  });

  it("Should call onClick", () => {
    const onClickSpy = jest.fn();
    const wrapper = shallow(
      <CharacterCard
        character={character}
        isFaved={false}
        onFaved={() => {}}
        onUnfaved={() => {}}
        onClick={onClickSpy}
      />
    );
    wrapper.simulate("click");
    expect(onClickSpy).toHaveBeenCalled();
  });

  it("Should call onFaved", () => {
    const onFaved = jest.fn();
    const wrapper = shallow(
      <CharacterCard
        character={character}
        isFaved={false}
        onFaved={onFaved}
        onUnfaved={() => {}}
        onClick={() => {}}
      />
    );
    const Star = wrapper.find(TopRightStar);
    Star.simulate("click", {
      stopPropagation: () => {},
    });
    expect(onFaved).toHaveBeenCalled();
  });

  it("Should call onUnfaved", () => {
    const onUnfaved = jest.fn();
    const wrapper = shallow(
      <CharacterCard
        character={character}
        onFaved={() => {}}
        onUnfaved={onUnfaved}
        onClick={() => {}}
        isFaved
      />
    );
    const Star = wrapper.find(TopRightStar);
    Star.simulate("click", {
      stopPropagation: () => {},
    });
    expect(onUnfaved).toHaveBeenCalled();
  });

  it("Should match snapshot", () => {
    const wrapper = shallow(
      <CharacterCard
        character={character}
        onFaved={() => {}}
        onUnfaved={() => {}}
        onClick={() => {}}
        isFaved
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
