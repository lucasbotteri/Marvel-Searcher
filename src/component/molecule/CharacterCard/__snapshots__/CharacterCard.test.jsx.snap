// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard Should match snapshot 1`] = `
<styled.section
  alt="Deathstrike (Ultimate)"
  onClick={[Function]}
>
  <styled.img
    src="someurl.com"
  />
  <Styled(Star)
    isSolid={true}
    onClick={[Function]}
  />
  <styled.h2>
     
    Deathstrike (Ultimate)
     
  </styled.h2>
</styled.section>
`;
